### ISSUE TITLE
EcsIsA runs very very slow

### ISSUE THREAD

**Issue by CodingJellyfish** [SanderMertens/flecs#843]
**Describe the bug**
It's hard to say if it's a bug, and I don't know whether it's my bug or not. But I added 1000 entities which each one represents a bezier line rendering, and the speed is 200 FPS without EcsIsA implementation while 50 FPS with it. 

System declaration:
```c
dodraw = ecs_system_init(ecs, &(ecs_system_desc_t) {
            .query = {
                .filter.terms = {
                    { ecs_id(draw_canvas_t),         .oper = EcsOptional },
                    { ecs_id(draw_color_t),          .oper = EcsOptional },
                    { ecs_id(draw_custom_t),         .oper = EcsOptional },
                    { ecs_id(draw_font_t),           .oper = EcsOptional },
                    { ecs_id(draw_shape_t),          .oper = EcsOptional },
                    { ecs_id(draw_sprite_t),         .oper = EcsOptional },
                    { ecs_id(draw_text_t),           .oper = EcsOptional },
                    { ecs_id(draw_texture_t),        .oper = EcsOptional },     
                    { ecs_id(transform_t),           .oper = EcsOptional },
                    { ecs_id(draw_mode_t),           .oper = EcsOptional },
                    { Draw },
                    { ecs_id(draw_priority_t), .inout = EcsIn }
                },
                .order_by_component = ecs_id(draw_priority_t),
                .order_by = cmp
            },
            .callback = do_draw
        });
```
Main with EcsIsA
```c
ecs_entity_t jay = ecs_new_id(ecs);
    ecs_add(ecs, jay, Draw);
    ecs_add_id(ecs, jay, EcsPrefab);
    ecs_set(ecs, jay, draw_color_t, { BEIGE });
    ecs_set(ecs, jay, draw_priority_t, { 3 });
    for (int i = 0; i < 1000; i++) {
        ecs_entity_t k = ecs_new_w_pair(ecs, EcsIsA, jay);
        ecs_set(ecs, k, draw_shape_t, {
            .type = DRAW_BEZIER,
            .thickness = 10.0f,
            .vertex = { {
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 }
            } }
        });
    }
```
Main without EcsIsA:
```c
    for (int i = 0; i < 1000; i++) {
        ecs_entity_t k = ecs_new_id(ecs);
        ecs_add(ecs, k, Draw);
        ecs_set(ecs, k, draw_color_t, { BEIGE });
        ecs_set(ecs, k, draw_priority_t, { 3 });
        ecs_set(ecs, k, draw_shape_t, {
            .type = DRAW_BEZIER,
            .thickness = 10.0f,
            .vertex = { {
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 },
                { rand() % 1000, rand() % 1000 }
            } }
        });
    }
```
Compare function: 
```c
static int cmp(ecs_entity_t e1, const void *v1, ecs_entity_t e2, const void *v2)
{
    const draw_priority_t *p1 = v1, *p2 = v2;
    return (p1->priority > p2->priority) - (p1->priority < p2->priority);
}
```
The only external library apart from FLECS is raylib.

**Comment by SanderMertens** [comment]
This is most likely caused by query instancing. See this part of the documentation:
https://www.flecs.dev/flecs/#/docs/Queries?id=instancing

Example:
https://github.com/SanderMertens/flecs/blob/master/examples/c/queries/instancing/src/main.c

TL;DR: by default queries/systems with shared components use a "safe" mode where the code iterating the query doesn't need to be aware of whether a component is shared or not. This safe mode returns entities one by one when the result has a shared component, which most likely is the cause of the slowdown you're seeing.

You can make a query instanced, which turns off the safety, and lets you iterate entities with shared components in bulk- but you will need to update your code to handle shared components. The manual/examples show how to do it.

**Comment by SanderMertens** [comment]
Closing this issue as instancing should solve it. Feel free to open another issue if after trying out instancing you still have performance issues!
